File: problems/basic/316/solution.py
# Accepted: 45ms (42.62%), 16.56MB (59.98%)

class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        lastOccurIndex: Dict[str, int] = {ch: i for i, ch in enumerate(s)}

        visited: Set[int] = set()
        ordered: List[str] = []

        for i in range(len(s)):
            ch = s[i]

            if ch in visited:
                continue

            # pop characters
            while len(ordered) > 0:
                top = ordered[-1]
                if i >= lastOccurIndex[top]: # last occurrence
                    break
                if ch > top: # in lexicographical order
                    break

                visited.remove(top)
                ordered.pop()

            # insert current character
            visited.add(ch)
            ordered.append(ch)

        return "".join(ordered)

File: problems/basic/316/solution.cpp
// Accepted: 3ms (54.61%), 9.52MB (5.04%)

class Solution {
public:
    string removeDuplicateLetters(string s) {
        unordered_map<char, int> lastOccurIndex;
        for (int i = 0; i < s.size(); ++i) {
            char ch = s[i];
            lastOccurIndex[ch] = i;
        }

        unordered_set<char> visited;
        vector<char> ordered;

        for (int i = 0; i < s.size(); ++i) {
            char ch = s[i];

            if (visited.count(ch) != 0) {
                continue;
            }

            // pop characters
            while (ordered.size() > 0) {
                char top = ordered.back();
                if (i >= lastOccurIndex[top]) { // last occurrence
                    break;
                }
                if (ch > top) { // in lexicographical order
                    break;
                }

                visited.erase(top);
                ordered.pop_back();
            }

            // insert current character
            visited.insert(ch);
            ordered.push_back(ch);
        }

        string orderedStr(ordered.begin(), ordered.end());
        return orderedStr;
    }
};
